pac4j-core/src/main/java/org/pac4j/core/profile/service/AbstractProfileService.java

```diff
@@ -230,9 +230,15 @@ protected U convertAttributesToProfile(final List<Map<String, Object>> listStora
             }
             final Object retrievedUsername = storageAttributes.get(getUsernameAttribute());
             if (retrievedUsername != null) {
-                profile.setId(ProfileHelper.sanitizeIdentifier(profile, retrievedUsername));
+                profile.addAttribute(USERNAME, retrievedUsername);
             } else {
-                profile.setId(username);
+                profile.addAttribute(USERNAME, username);
+            }
+            final Object retrievedId = storageAttributes.get(getIdAttribute());
+            if (retrievedId != null) {
+                profile.setId(ProfileHelper.sanitizeIdentifier(profile, retrievedId));
+            } else {
```

Comment:
Else clause unnecessary.

Reply:
Well, most Technical exceptions are stuff that "should" not happen and are sanity check. But I can remove it.